File: li5/LastPageViewController.m
//
//  LastPageViewController.m
//  li5
//
//  Created by Leandro Fournier on 4/27/16.
//  Copyright © 2016 Li5, Inc. All rights reserved.
//

#import "ExploreDynamicInteractor.h"
#import "LastPageViewController.h"
#import "Li5Constants.h"
#import "Li5VolumeView.h"
#import "UserProfileDynamicInteractor.h"

@interface LastPageViewController ()
{
    UIPanGestureRecognizer *profilePanGestureRecognizer;
    UIPanGestureRecognizer *searchPanGestureRecognizer;
    id<UserProfileViewControllerPanTargetDelegate> profileInteractor;
    id<ExploreViewControllerPanTargetDelegate> searchInteractor;
    
    id __playerEndObserver;
    BOOL __hasAppeared;
    id __showPlayerEndObserver;
}

@property (weak, nonatomic) IBOutlet UIView *staticView;
@property (weak, nonatomic) IBOutlet UIView *videoView;
@property (weak, nonatomic) IBOutlet UILabel *closeMessage;
@property (weak, nonatomic) IBOutlet UIView *swipeDownView;
@property (weak, nonatomic) IBOutlet UIButton *turnOnNotifications;
@property (weak, nonatomic) IBOutlet UIImageView *dropDownArrow;

@property (weak, nonatomic) IBOutlet UIImageView *popcorn1;
@property (weak, nonatomic) IBOutlet UIImageView *popcorn2;
@property (weak, nonatomic) IBOutlet UIImageView *popcornFloor;
@property (weak, nonatomic) IBOutlet UIImageView *popcorn3;
@property (weak, nonatomic) IBOutlet UIImageView *popcorn4;

@property (nonatomic, strong) BCPlayer *player;
@property (nonatomic, strong) BCPlayerLayer *playerLayer;
@property (nonatomic, strong) AVAudioPlayer* audioPlayer;

@property (weak, nonatomic) IBOutlet UIView *endOfShowView;
@property (nonatomic, strong) AVPlayer *showPlayer;
@property (weak, nonatomic) IBOutlet UIView *endOfShowVideoView;
@property (weak, nonatomic) IBOutlet UIImageView *showLogo;

@end

@implementation LastPageViewController

@synthesize product;

#pragma mark - Init

- (instancetype)init
{
    self = [super init];
    if (self)
    {
        [self initialize];
    }
    return self;
}

- (instancetype)initWithCoder:(NSCoder *)aDecoder
{
    self = [super initWithCoder:aDecoder];
    if (self)
    {
        [self initialize];
    }
    return self;
}

- (void)initialize
{
    
}

#pragma mark - Public Methods

- (void)setLastVideoURL:(EndOfPrimeTime *)endOfPrimeTime
{
    _lastVideoURL = endOfPrimeTime;
    
    _player = [[BCPlayer alloc] initWithUrl:[NSURL URLWithString:endOfPrimeTime.url] bufferInSeconds:50.0 priority:BCPriorityPlay delegate:self];
    _playerLayer = [[BCPlayerLayer alloc] initWithPlayer:_player andFrame:[UIScreen mainScreen].bounds previewImageRequired:YES];
    
    __weak typeof(self) welf = self;
    [[NSOperationQueue mainQueue] addOperationWithBlock:^{
        NSData *posterData = [[NSData alloc] initWithBase64EncodedString:endOfPrimeTime.poster options:0];
        UIImage *posterImage = [UIImage imageWithData:posterData];
        UIImageView *posterImageView = [[UIImageView alloc] initWithImage:posterImage];
        posterImageView.frame = welf.view.bounds;
        [welf.videoView insertSubview:posterImageView atIndex:0];
    }];
}

#pragma mark - UI Setup

- (void)viewDidLoad
{
    DDLogVerbose(@"");
    [super viewDidLoad];
    // Do any additional setup after loading the view from its nib.
    
    self.staticView.hidden = YES;
    
    _playerLayer.frame = self.view.bounds;
    _playerLayer.videoGravity = AVLayerVideoGravityResizeAspectFill;
    
    [self.videoView.layer addSublayer:_playerLayer];

//    NSString *goToExploreSoundURL = [[NSBundle mainBundle] pathForResource:@"go_to_explore" ofType:@"mp3"];
//    _audioPlayer = [[AVAudioPlayer alloc] initWithContentsOfURL:[NSURL fileURLWithPath:goToExploreSoundURL] error:nil];
//    [_audioPlayer setNumberOfLoops:0];
//    [_audioPlayer prepareToPlay];

    NSNotificationCenter *notificationCenter = [NSNotificationCenter defaultCenter];
    [notificationCenter postNotificationName:kPrimeTimeLoaded object:nil];
    
    [self setupGestureRecognizers];
    
    [self.view addSubview:[[Li5VolumeView alloc] initWithFrame:CGRectMake(0, 0, self.view.frame.size.width, 5.0)]];
    
#if FULL_VERSION
    self.closeMessage.text = NSLocalizedString(@"SWIPE DOWN TO EXPLORE MORE",nil);
#endif
    
    [self updateNotificationsViews];
    
#ifdef EMBED
    self.showLogo.image = [self.showLogo.image imageWithRenderingMode:UIImageRenderingModeAlwaysTemplate];
    self.showLogo.tintColor = [UIColor li5_whiteColor];
    
    NSURL *videoURL = [[NSBundle mainBundle] URLForResource:@"end_of_show" withExtension:@".mp4"];
    self.showPlayer = [[AVPlayer alloc] initWithURL:videoURL];
    self.showPlayer.actionAtItemEnd = AVPlayerActionAtItemEndNone;
    self.showPlayer.muted = TRUE;
    
    AVPlayerLayer *videoLayer = [AVPlayerLayer playerLayerWithPlayer:self.showPlayer];
    videoLayer.frame = self.view.bounds;
    videoLayer.videoGravity = AVLayerVideoGravityResizeAspectFill;
    
    [self.endOfShowVideoView.layer addSublayer:videoLayer];
    
    self.staticView.hidden = YES;
    self.swipeDownView.hidden = YES;
#else
    self.endOfShowView.hidden = YES;
#endif
}

- (void)viewWillAppear:(BOOL)animated {
    DDLogVerbose(@"");
    [super viewWillAppear:animated];
    
    [self updateNotificationsViews];
}

- (void)viewDidAppear:(BOOL)animated
{
    DDLogVerbose(@"");
    [super viewDidAppear:animated];
    
    __hasAppeared = YES;
    
    [self readyToPlay];
    [self setupObservers];
}

- (void)viewDidDisappear:(BOOL)animated
{
    DDLogVerbose(@"");
    [super viewDidDisappear:animated];
 
    __hasAppeared = NO;
    
    [self.player pause];
    [self removeObservers];
    
    self.staticView.hidden = (self.player!=nil);
}

- (void)updateNotificationsViews {
    DDLogVerbose(@"");
    if ([self notificationsEnabled]) {
        self.turnOnNotifications.hidden = YES;
        self.dropDownArrow.hidden = YES;
        for (UIView *v in @[_popcorn1, _popcorn2, _popcorn3, _popcorn4, _popcornFloor]) {
            v.hidden = NO;
        }
    } else {
        self.turnOnNotifications.hidden = NO;
        self.dropDownArrow.hidden = NO;
        for (UIView *v in @[_popcorn1, _popcorn2, _popcorn3, _popcorn4, _popcornFloor]) {
            v.hidden = YES;
        }
        [self __bounce:self.dropDownArrow];
    }
    
    [UIView animateWithDuration:0.25 animations:^{
        [self.view layoutIfNeeded];
    }];
}

- (BOOL)notificationsEnabled {
    return YES;
}

- (void)presentSwipeDownViewIfNeeded
{
    DDLogVerbose(@"");
    [self removeObservers];
    
#ifndef EMBED
#if FULL_VERSION
    NSUserDefaults *userDefaults = [NSUserDefaults standardUserDefaults];
    if (![userDefaults boolForKey:kLi5SwipeDownExplainerViewPresented])
    {
        [_audioPlayer play];
        self.swipeDownView.hidden = NO;
        [self hideVideo];
    }
    else
    {
        [_audioPlayer play];
        [self hideVideo];
        [self.view bringSubviewToFront:self.staticView];
    }
#else
    [_audioPlayer play];
    [self hideVideo];
    [self.view bringSubviewToFront:self.staticView];
#endif
#else
    [_audioPlayer play];
    [self hideVideo];
    [self.showPlayer play];
    [self setupObservers];
#endif
}

- (void)__bounce:(UIView *)view
{
    CAKeyframeAnimation *trans = [CAKeyframeAnimation animationWithKeyPath:@"position.y"];
    trans.values = @[@(0),@(5),@(-2.0),@(3),@(0)];
    trans.keyTimes = @[@(0.0),@(0.35),@(0.70),@(0.90),@(1)];
    trans.timingFunction = [CAMediaTimingFunction functionWithControlPoints:.5 :1.8 :1 :1];
    trans.duration = 2.0;
    trans.additive = YES;
    trans.repeatCount = INFINITY;
    trans.beginTime = CACurrentMediaTime() + 2.0;
    trans.removedOnCompletion = NO;
    trans.fillMode = kCAFillModeForwards;
    [view.layer addAnimation:trans forKey:@"bouncing"];
}


- (void)hideVideo
{
    if (_player)
    {
        self.staticView.hidden = NO;
        self.videoView.hidden = YES;
        [_playerLayer removeFromSuperlayer];
        _playerLayer = nil;
        _player = nil;
    }
}

- (IBAction)doTurnOnNotifications:(id)sender {
    
}

- (void)replayEndOfShow {
    [self.showPlayer seekToTime:kCMTimeZero];
    [self.showPlayer play];
}

#pragma mark - Observers

- (void)setupObservers
{
    DDLogVerbose(@"");
    if (!__playerEndObserver && _player)
    {
        __weak typeof(self) welf = self;
        __playerEndObserver = [[NSNotificationCenter defaultCenter] addObserverForName:AVPlayerItemDidPlayToEndTimeNotification object:self.player.currentItem queue:NSOperationQueuePriorityNormal usingBlock:^(NSNotification *_Nonnull note) {
            [welf presentSwipeDownViewIfNeeded];
        }];
    }
    
    if (!__showPlayerEndObserver && _showPlayer) {
        __weak typeof(self) welf = self;
        __showPlayerEndObserver = [[NSNotificationCenter defaultCenter] addObserverForName:AVPlayerItemDidPlayToEndTimeNotification object:self.showPlayer.currentItem queue:NSOperationQueuePriorityNormal usingBlock:^(NSNotification *_Nonnull note) {
            [welf replayEndOfShow];
        }];
    }
    
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(updateNotificationsViews) name:kUserSettingsUpdated object:nil];
}

- (void)removeObservers
{
    DDLogVerbose(@"");
    if (__playerEndObserver)
    {
        [[NSNotificationCenter defaultCenter] removeObserver:__playerEndObserver];
        __playerEndObserver = nil;
    }
    
    if (__showPlayerEndObserver) {
        [[NSNotificationCenter defaultCenter] removeObserver:__showPlayerEndObserver];
        __showPlayerEndObserver = nil;
    }
    
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

#pragma mark - Player

- (void)readyToPlay
{
    DDLogVerbose(@"");
//    if (self.player.status == AVPlayerStatusReadyToPlay)
//    {
        if (__hasAppeared && self.player != nil)
        {
            [self.player play];
        }
        else
        {
            self.staticView.hidden = NO;
//            [self presentSwipeDownViewIfNeeded];
        }
//    }
}

- (void)bufferEmpty
{
    DDLogVerbose(@"");
}

- (void)bufferReady
{
    DDLogVerbose(@"");
}

- (void)failToLoadItem:(NSError *)error
{
    DDLogError(@"%@",error.description);
    [[CrashlyticsLogger sharedInstance] logError:error userInfo:nil];
}

- (void)networkFail:(NSError *)error
{
    DDLogError(@"%@",error.description);
    [[CrashlyticsLogger sharedInstance] logError:error userInfo:nil];
}

#pragma mark - Gesture Recognizers

- (void)userDidPan:(UIPanGestureRecognizer *)recognizer {
    if (self.videoView.hidden) {
        if (recognizer.state == UIGestureRecognizerStateBegan) {
            if (!self.swipeDownView.hidden) {
                self.swipeDownView.hidden = YES;
                NSUserDefaults *userDefaults = [NSUserDefaults standardUserDefaults];
                [userDefaults setBool:TRUE forKey:kLi5SwipeDownExplainerViewPresented];
                [self.view bringSubviewToFront:self.staticView];
            }
        }
        
        [searchInteractor userDidPan:recognizer];
    }
}

- (void)setupGestureRecognizers
{
    DDLogVerbose(@"");
//    //User Profile Gesture Recognizer - Swipe Down from 0-100px
//    profileInteractor = [[UserProfileDynamicInteractor alloc] initWithParentViewController:self];
//    profilePanGestureRecognizer = [[UIPanGestureRecognizer alloc] initWithTarget:profileInteractor action:@selector(userDidPan:)];
//    [profilePanGestureRecognizer setDelegate:self];
//    [self.view addGestureRecognizer:profilePanGestureRecognizer];
    
#if FULL_VERSION
    //Search Products Gesture Recognizer - Swipe Down from below 100px
    searchPanGestureRecognizer = [[UIPanGestureRecognizer alloc] initWithTarget:self action:@selector(userDidPan:)];
    [searchPanGestureRecognizer setDelegate:self];
    [self.view addGestureRecognizer:searchPanGestureRecognizer];
#endif
    
}

- (BOOL)gestureRecognizer:(UIGestureRecognizer *)gestureRecognizer shouldRecognizeSimultaneouslyWithGestureRecognizer:(UIGestureRecognizer *)otherGestureRecognizer
{
    return NO;
}

- (BOOL)gestureRecognizerShouldBegin:(UIGestureRecognizer *)gestureRecognizer
{
    CGPoint touch = [gestureRecognizer locationInView:gestureRecognizer.view];
    if (gestureRecognizer == profilePanGestureRecognizer)
    {
        CGPoint velocity = [(UIPanGestureRecognizer*)gestureRecognizer velocityInView:gestureRecognizer.view];
        return (touch.y < 150) && (velocity.y > 0);
    }
    else if (gestureRecognizer == searchPanGestureRecognizer)
    {
        CGPoint velocity = [(UIPanGestureRecognizer*)gestureRecognizer velocityInView:gestureRecognizer.view];
        double degree = atan(velocity.y/velocity.x) * 180 / M_PI;
        return (touch.y >= 150) && (fabs(degree) > 70.0) && (velocity.y > 0);
    }
    return false;
}

- (BOOL)gestureRecognizer:(UIGestureRecognizer *)gestureRecognizer
shouldRequireFailureOfGestureRecognizer:(UIGestureRecognizer *)otherGestureRecognizer
{
    if ([[gestureRecognizer view] isKindOfClass:[UIScrollView class]])
    {
        if (otherGestureRecognizer == profilePanGestureRecognizer || otherGestureRecognizer == searchPanGestureRecognizer)
        {
            return YES;
        }
    }
    return (gestureRecognizer == profilePanGestureRecognizer &&
            (otherGestureRecognizer == searchPanGestureRecognizer));
}

#pragma mark - OS Actions

- (void)dealloc
{
    DDLogDebug(@"%p",self);
    [self removeObservers];
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

@end

File: Li5SDK/LastPageViewController.swift
//
//  LastPageViewController.swift
//  li5
//
//  Created by Sergio Daniel L. García on 6/10/17.
//  Copyright © 2017 Li5, Inc. All rights reserved.
//

import Foundation
import Crashlytics


protocol LastPageViewControllerProtocol {
    
    var content: EndOfPrimeTime { get set }
    
}


class LastPageViewController : PaginatorViewController, LastPageViewControllerProtocol {
    
    var isBeingDisplayed = false
    
    var content: EndOfPrimeTime
    var player: AVPlayer?
    var playerLayer: AVPlayerLayer?
    var showPlayer: AVPlayer?
    
    var playbackEndObserver: Any?
    var showPlaybackEndObserver: Any?
    
    var profilePanGestureRecognizer: UIPanGestureRecognizer?
    
    @IBOutlet weak var videoView: UIView!
    @IBOutlet weak var staticView: UIView!
    @IBOutlet weak var showLogo: UIImageView!
    @IBOutlet weak var endOfShowVideoView: UIView!
    @IBOutlet weak var swipeDownView: UIView!
    @IBOutlet weak var turnOnNotifications: UIButton!
    @IBOutlet weak var dropDownArrow: UIImageView!
    @IBOutlet weak var closeMessage: UILabel!
    @IBOutlet var popcorns: [UIImageView]!
    @IBOutlet weak var endOfShowView: UIView!
    
    private var notificationsEnabled: Bool {
        return true
    }
    
    var lastVideoUrl: EndOfPrimeTime? {
        didSet {
            guard let lastVideoUrl = lastVideoUrl else {
                log.error("This doesn't make any sense. lastVideoUrl was set but it wasn't at the same time? It's nil!")
                return
            }
            
            if let url = Foundation.URL(string: lastVideoUrl.url) {
                player = AVPlayer(url: url)
                playerLayer = AVPlayerLayer(player: player)
                
                OperationQueue.main.addOperation { [weak self] in
                    guard let this = self else {
                        log.warning("Lost reference of LastPageViewController self while in queue operation closure")
                        return
                    }
                    
                    if let lastVideoUrl = this.lastVideoUrl {
                        guard let posterData = Data(base64Encoded: lastVideoUrl.poster) else {
                            log.error("Poster data for LastPageViewController could not be converted from base64")
                            return
                        }
                        
                        guard let posterImage = UIImage(data: posterData) else {
                            log.error("Poster data for LastPageViewController could not be converted to Image")
                            return
                        }
                        
                        let posterImageView = UIImageView(image: posterImage)
                        posterImageView.frame = this.view.bounds
                        this.videoView.insertSubview(posterImageView, at: 0)
                    }
                }
            } else {
                log.error("Coudln't create player because URL string could not be parsed for LastPageViewController")
            }
            
        }
    }
    
    var endOfShow: EndOfShow?
    
    init(content: EndOfPrimeTime) {
        self.content = content
        super.init()
        self.reset()
    }
    
    required public init?(coder aDecoder: NSCoder) {
        self.content = EndOfPrimeTime()
        super.init(coder: aDecoder)
        self.reset()
    }
    
    public required init(withProduct product: ProductModel) {
        fatalError("init(withProduct:pageIndex:andContext:) has not been implemented")
    }
    
    static func instance(withEOP eop: EndOfPrimeTime) -> LastPageViewController {
        
        let storyboard = UIStoryboard(name: "DiscoverViews",
                                      bundle: Bundle(for: TrailerViewController.self))
        
        let vc = storyboard.instantiateViewController(withIdentifier: "LastPageView")
            as? LastPageViewController
        
        if vc == nil {
            log.error("Failed to cast ViewController from storyboard to LastPageViewController")
        }
        
        let viewController = vc!
        
        log.verbose("Initializing new LastPageVC")
        
        viewController.content = eop
        
        return viewController
    }
    
    // MARK: - LIFECYCLE
    
    override func viewDidLoad() {
        log.verbose("LastPageViewController did load")
        super.viewDidLoad()
        
        staticView.isHidden = true
        
        playerLayer?.frame = view.bounds
        playerLayer?.videoGravity = AVLayerVideoGravityResizeAspectFill
        
        if let playerLayer = playerLayer {
            videoView.layer.addSublayer(playerLayer)
        }
        
        view.addSubview(Li5VolumeView(frame: CGRect(x: 0, y: 0, width: view.frame.size.width, height: 5)))
        
        updateNotificationsView()
        
        showLogo.image = showLogo.image?.withRenderingMode(UIImageRenderingMode.alwaysTemplate)
        showLogo.tintColor = UIColor.li5_white()
        
        if endOfShow?.url != nil, let videoUrl = Foundation.URL(string: endOfShow!.url) {
            showPlayer = AVPlayer(url: videoUrl)
            showPlayer?.actionAtItemEnd = .none
            showPlayer?.isMuted = true
            
            let videoLayer = AVPlayerLayer(player: showPlayer)
            
            videoLayer.frame = view.bounds
            videoLayer.videoGravity = AVLayerVideoGravityResizeAspectFill
            
            endOfShowVideoView.layer.addSublayer(videoLayer)
            staticView.isHidden = true
            swipeDownView.isHidden = true
        } else {
            log.error("Could not get URL for resource end_of_show.mp4 in LastPageViewController")
        }
    }
    
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        updateNotificationsView()
    }
    
    
    override func viewDidAppear(_ animated: Bool) {
        log.verbose("LastPageViewController did appear")
        super.viewDidAppear(animated)
        isBeingDisplayed = true
        play()
        setupObservers()
    }
    
    
    override func viewDidDisappear(_ animated: Bool) {
        log.verbose("LastPageViewController did disappear")
        super.viewDidDisappear(animated)
        isBeingDisplayed = false
        player?.pause()
        showPlayer?.pause()
        clearObservers()
        staticView.isHidden = player != nil
    }
    
    @objc private func updateNotificationsView() {
        log.verbose("Updating notifications view")
        
        if notificationsEnabled {
            turnOnNotifications.isHidden = true
            dropDownArrow.isHidden = true
            
            for popcorn in popcorns {
                popcorn.isHidden = false
            }
        } else {
            turnOnNotifications.isHidden = false
            dropDownArrow.isHidden = false
            
            for popcorn in popcorns {
                popcorn.isHidden = true
            }
            
            bounce(view: dropDownArrow)
        }
        
        UIView.animate(withDuration: 0.25) { [unowned self] in
            self.view.layoutIfNeeded()
        }
    }
    
    fileprivate func play() {
        if lastVideoUrl != nil, isBeingDisplayed, let player = player {
            player.seek(to: kCMTimeZero)
            player.play()
        } else {
            replayEndOfShow()
        }
    }
    
    private func setupObservers() {
        log.verbose("Setting up observers for LastPageViewController")
        
        if playbackEndObserver == nil && player != nil {
            playbackEndObserver = NotificationCenter.default.addObserver(forName: .AVPlayerItemDidPlayToEndTime, object: player?.currentItem, queue: nil, using: { [weak self] (notification) in
                self?.presentSwipeDownViewIfNeeded()
            })
        }
        
        if showPlaybackEndObserver == nil && showPlayer != nil {
            showPlaybackEndObserver = NotificationCenter.default.addObserver(forName: .AVPlayerItemDidPlayToEndTime, object: showPlayer?.currentItem, queue: nil, using: { [weak self] (notification) in
                self?.replayEndOfShow()
            })
        }
        
        NotificationCenter.default.addObserver(self, selector: #selector(updateNotificationsView), name: NSNotification.Name(rawValue: "kUserSettingsUpdated"), object: nil)
    }
    
    
    private func clearObservers() {
        log.verbose("Clearing observers for LastPageViewController")
        
        if let observer = playbackEndObserver {
            NotificationCenter.default.removeObserver(observer)
            playbackEndObserver = nil
        }
        
        if let observer = showPlaybackEndObserver {
            NotificationCenter.default.removeObserver(observer)
            showPlaybackEndObserver = nil
        }
        
        NotificationCenter.default.removeObserver(self)
    }
    
    
    private func bounce(view: UIView) {
        let transition = CAKeyframeAnimation(keyPath: "position.y")
        transition.values = [0, 5, -2.0, 3, 0]
        transition.keyTimes = [0.0, 0.35, 0.7, 0.9, 1]
        transition.timingFunction = CAMediaTimingFunction(controlPoints: 0.5, 1.8, 1, 1)
        transition.duration = 2.0
        transition.isAdditive = true
        transition.repeatCount = Float.infinity
        transition.beginTime = CACurrentMediaTime() + 2.0
        transition.isRemovedOnCompletion = false
        transition.fillMode = kCAFillModeForwards
        view.layer.add(transition, forKey: "bouncing")
    }
    
    
    private func presentSwipeDownViewIfNeeded() {
        log.verbose("Running presentSwipeDownViewIfNeeded")
        clearObservers()
        
        hideVideo()
        showPlayer?.isMuted = false
        showPlayer?.play()
        setupObservers()
    }
    
    
    private func hideVideo() {
        if player != nil {
            videoView.isHidden = true
            playerLayer?.removeFromSuperlayer()
            playerLayer = nil
            player = nil
        }
    }
    
    
    func replayEndOfPrimeTime() {
        player?.seek(to: kCMTimeZero)
        player?.play()
    }
    
    
    private func replayEndOfShow() {
        showPlayer?.seek(to: kCMTimeZero)
        showPlayer?.play()
    }
    
    
    // MARK: GESTURE RECOGNIZERS
    
    func userDidPan(_ recognizer: UIPanGestureRecognizer) {
        if videoView.isHidden {
            if recognizer.state == .began {
                if swipeDownView.isHidden == false {
                    swipeDownView.isHidden = true
                    UserDefaults.standard.set(true, forKey: "Li5SwipeDownExplainerViewPresented")
                    view.bringSubview(toFront: staticView)
                }
            }
            
            //    [searchInteractor userDidPan:recognizer];
        }
    }

    
    func gestureRecognizer(_ gestureRecognizer: UIGestureRecognizer, shouldRecognizeSimultaneouslyWith otherGestureRecognizer: UIGestureRecognizer) -> Bool {
        return false
    }
    
    
    func gestureRecognizerShouldBegin(_ gestureRecognizer: UIGestureRecognizer) -> Bool {
        let touch = gestureRecognizer.location(in: gestureRecognizer.view)
        
        if gestureRecognizer == profilePanGestureRecognizer {
            if let gestureRecognizer = gestureRecognizer as? UIPanGestureRecognizer {
                let velocity = gestureRecognizer.velocity(in: gestureRecognizer.view)
                return touch.y < 150 && velocity.y > 0
            }
        }
        
        return false
    }
    
    
    func gestureRecognizer(_ gestureRecognizer: UIGestureRecognizer, shouldRequireFailureOf otherGestureRecognizer: UIGestureRecognizer) -> Bool {
        
        if gestureRecognizer.view is UIScrollView {
            return otherGestureRecognizer == profilePanGestureRecognizer
        }
        
        return gestureRecognizer == profilePanGestureRecognizer
    }
    
    
}


extension LastPageViewController : UIGestureRecognizerDelegate {
    
    
    
}
